File: app/api/v1/produkter.py
"""Product API endpoints."""
from typing import List, Optional, Dict, Any
from sqlalchemy import select, or_, func, text, String
from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import APIRouter, Depends, HTTPException, Query, Body
from pydantic import BaseModel

from app.api.deps import get_db, get_current_user
from app.domain.entities.user import User
from app.models.produkter import Produkter as ProdukterModel
from app.models.matinfo_products import MatinfoProduct, MatinfoAllergen, MatinfoNutrient
from app.schemas.produkter import Produkter, ProdukterCreate, ProdukterUpdate
from sqlalchemy.orm import selectinload

router = APIRouter()


class ProdukterListResponse(BaseModel):
    """Paginated response for products list."""
    items: List[Produkter]
    total: int
    page: int
    page_size: int
    total_pages: int


@router.get("/", response_model=ProdukterListResponse)
async def get_produkter(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=10000),
    aktiv: Optional[bool] = Query(None, description="Filter by active status"),
    kategori: Optional[int] = Query(None, description="Filter by category ID"),
    sok: Optional[str] = Query(None, description="Search in product name"),
    rett_komponent: Optional[bool] = Query(None, description="Filter by dish component status"),
    has_gtin: Optional[bool] = Query(None, description="Filter by GTIN presence: true=has GTIN, false=missing GTIN"),
    leverandor_ids: Optional[str] = Query(None, description="Comma-separated list of leverandor IDs to filter by"),
    sort_by: Optional[str] = Query(None, description="Sort by field (produktid, produktnavn, pris, ean_kode)"),
    sort_order: Optional[str] = Query("asc", description="Sort order: asc or desc"),
) -> ProdukterListResponse:
    """Get all products with pagination."""
    # Build base query
    base_query = select(ProdukterModel)

    # Filter by active status
    if aktiv is not None:
        if aktiv:
            base_query = base_query.where(
                or_(ProdukterModel.utgatt == False, ProdukterModel.utgatt == None)
            )
        else:
            base_query = base_query.where(ProdukterModel.utgatt == True)

    # Filter by category
    if kategori:
        base_query = base_query.where(ProdukterModel.kategoriid == kategori)

    # Filter by leverandor IDs (comma-separated list)
    if leverandor_ids:
        try:
            ids = [int(id.strip()) for id in leverandor_ids.split(",") if id.strip()]
            if ids:
                base_query = base_query.where(ProdukterModel.levrandorid.in_(ids))
        except ValueError:
            pass  # Ignore invalid IDs

    # Filter by rett_komponent (dish component)
    if rett_komponent is not None:
        base_query = base_query.where(ProdukterModel.rett_komponent == rett_komponent)

    # Filter by GTIN presence
    if has_gtin is not None:
        if has_gtin:
            base_query = base_query.where(
                or_(
                    ProdukterModel.ean_kode != None,
                    ProdukterModel.ean_kode != ""
                )
            )
        else:
            base_query = base_query.where(
                or_(
                    ProdukterModel.ean_kode == None,
                    ProdukterModel.ean_kode == ""
                )
            )

    # Search by name, ID, or codes
    if sok:
        search_term = f"%{sok}%"
        base_query = base_query.where(
            or_(
                ProdukterModel.produktnavn.ilike(search_term),
                ProdukterModel.visningsnavn.ilike(search_term),
                ProdukterModel.leverandorsproduktnr.ilike(search_term),
                ProdukterModel.ean_kode.ilike(search_term),
                func.cast(ProdukterModel.produktid, String).ilike(search_term)
            )
        )

    # Apply sorting
    if sort_by:
        sort_column = getattr(ProdukterModel, sort_by, None)
        if sort_column is not None:
            if sort_order == "desc":
                base_query = base_query.order_by(sort_column.desc())
            else:
                base_query = base_query.order_by(sort_column.asc())
        else:
            # Default sort by produktid if invalid sort field
            base_query = base_query.order_by(ProdukterModel.produktid.asc())
    else:
        # Default sort by produktid
        base_query = base_query.order_by(ProdukterModel.produktid.asc())

    # Get total count
    count_query = select(func.count()).select_from(base_query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0

    # Get paginated results
    query = base_query.offset(skip).limit(limit)
    result = await db.execute(query)
    produkter = result.scalars().all()

    # Calculate pagination
    page = (skip // limit) + 1 if limit > 0 else 1
    total_pages = (total + limit - 1) // limit if limit > 0 else 0

    return ProdukterListResponse(
        items=produkter,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages
    )


@router.get("/{produkt_id}", response_model=Produkter)
async def get_produkt(
    produkt_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Produkter:
    """Get a product by ID."""
    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid == produkt_id)
    )
    produkt = result.scalar_one_or_none()

    if not produkt:
        raise HTTPException(status_code=404, detail="Produkt ikke funnet")

    return produkt


@router.post("/by-ids", response_model=List[Produkter])
async def get_produkter_by_ids(
    produkt_ids: List[int] = Body(..., description="List of product IDs"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> List[Produkter]:
    """Get multiple products by their IDs."""
    if not produkt_ids:
        return []

    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid.in_(produkt_ids))
    )
    return result.scalars().all()


@router.post("/", response_model=Produkter)
async def create_produkt(
    produkt_data: ProdukterCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Produkter:
    """Create a new product."""
    produkt = ProdukterModel(**produkt_data.model_dump())
    db.add(produkt)
    await db.commit()
    await db.refresh(produkt)
    return produkt


@router.put("/{produkt_id}", response_model=Produkter)
async def update_produkt(
    produkt_id: int,
    produkt_data: ProdukterUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Produkter:
    """Update a product."""
    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid == produkt_id)
    )
    produkt = result.scalar_one_or_none()
    
    if not produkt:
        raise HTTPException(status_code=404, detail="Produkt ikke funnet")
    
    update_data = produkt_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(produkt, field, value)
    
    await db.commit()
    await db.refresh(produkt)
    return produkt


@router.delete("/{produkt_id}")
async def delete_produkt(
    produkt_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Delete a product (soft delete by setting utgatt=True)."""
    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid == produkt_id)
    )
    produkt = result.scalar_one_or_none()
    
    if not produkt:
        raise HTTPException(status_code=404, detail="Produkt ikke funnet")
    
    produkt.utgatt = True
    await db.commit()

    return {"message": "Produkt slettet"}


# Pydantic models for new endpoints
class MatinfoAllergenInfo(BaseModel):
    """Allergen informasjon fra Matinfo."""
    code: str
    name: str
    level: str  # FREE_FROM, CONTAINS, MAY_CONTAIN

    class Config:
        from_attributes = True


class MatinfoNutrientInfo(BaseModel):
    """Nutrient informasjon fra Matinfo."""
    code: str
    name: str
    measurement: Optional[float] = None
    measurement_precision: Optional[str] = None
    measurement_type: Optional[str] = None

    class Config:
        from_attributes = True


class MatinfoSearchResult(BaseModel):
    """Matinfo søkeresultat."""
    id: int
    gtin: str
    name: str
    brand: Optional[str] = None
    ingredients: Optional[str] = None
    similarity_score: Optional[float] = None
    allergens: List["MatinfoAllergenInfo"] = []
    nutrients: List["MatinfoNutrientInfo"] = []


class GtinUpdateRequest(BaseModel):
    """Request for å oppdatere GTIN."""
    produktid: int
    gtin: str


class BulkGtinUpdateRequest(BaseModel):
    """Request for masse-oppdatering av GTIN."""
    updates: List[GtinUpdateRequest]


@router.get("/matinfo/search", response_model=List[MatinfoSearchResult])
async def search_matinfo_products(
    query: str = Query(..., min_length=2, description="Søketekst (produktnavn, merke, ingredienser)"),
    limit: int = Query(20, ge=1, le=100, description="Maks antall resultater"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> List[MatinfoSearchResult]:
    """
    Fuzzy-søk i Matinfo-produkter.

    Søker i:
    - Produktnavn
    - Merkenavn
    - Ingredienser

    Bruker PostgreSQL trigram similarity for fuzzy matching.
    Returnerer også allergen og nutrition data.
    """
    # Normalize søketekst
    search_term = query.strip().lower()

    # PostgreSQL trigram similarity søk for å få IDs
    # Krever extension: CREATE EXTENSION IF NOT EXISTS pg_trgm;
    sql = text("""
        SELECT
            id,
            GREATEST(
                similarity(LOWER(name), :search),
                similarity(LOWER(COALESCE(brand_name, '')), :search),
                similarity(LOWER(COALESCE(ingredient_statement, '')), :search)
            ) as similarity_score
        FROM matinfo_products
        WHERE
            LOWER(name) % :search
            OR LOWER(COALESCE(brand_name, '')) % :search
            OR LOWER(COALESCE(ingredient_statement, '')) % :search
            OR LOWER(name) LIKE :like_search
            OR LOWER(COALESCE(brand_name, '')) LIKE :like_search
        ORDER BY similarity_score DESC
        LIMIT :limit
    """)

    result = await db.execute(
        sql,
        {
            "search": search_term,
            "like_search": f"%{search_term}%",
            "limit": limit
        }
    )

    rows = result.fetchall()
    product_ids = [row.id for row in rows]
    similarity_map = {row.id: row.similarity_score for row in rows}

    # Hent produkter med allergen og nutrition data
    products_result = await db.execute(
        select(MatinfoProduct)
        .options(selectinload(MatinfoProduct.allergens), selectinload(MatinfoProduct.nutrients))
        .where(MatinfoMatinfoProduct.id.in_(product_ids))
    )
    products = products_result.scalars().all()

    # Bygg respons med sortering basert på similarity
    products_map = {p.id: p for p in products}

    return [
        MatinfoSearchResult(
            id=pid,
            gtin=products_map[pid].gtin,
            name=products_map[pid].name,
            brand=products_map[pid].brand_name,
            ingredients=products_map[pid].ingredient_statement,
            similarity_score=round(float(similarity_map[pid]), 3) if similarity_map[pid] else None,
            allergens=[
                MatinfoAllergenInfo(
                    code=a.code,
                    name=a.name,
                    level=a.level
                ) for a in products_map[pid].allergens
            ],
            nutrients=[
                MatinfoNutrientInfo(
                    code=n.code,
                    name=n.name,
                    measurement=float(n.measurement) if n.measurement else None,
                    measurement_precision=n.measurement_precision,
                    measurement_type=n.measurement_type
                ) for n in products_map[pid].nutrients
            ]
        )
        for pid in product_ids if pid in products_map
    ]


@router.patch("/{produkt_id}/gtin")
async def update_produkt_gtin(
    produkt_id: int,
    gtin: str = Body(..., embed=True, description="Ny GTIN/EAN-kode"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Dict[str, Any]:
    """
    Oppdater GTIN/EAN-kode for et produkt.

    Renser automatisk GTIN (fjerner bindestreker, validerer lengde).
    """
    # Hent produkt
    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid == produkt_id)
    )
    produkt = result.scalar_one_or_none()

    if not produkt:
        raise HTTPException(status_code=404, detail="Produkt ikke funnet")

    # Rens GTIN
    clean_gtin = gtin.strip().replace("-", "").replace(" ", "")

    # Valider GTIN-lengde (8, 12, 13, eller 14 siffer)
    if clean_gtin and len(clean_gtin) not in [8, 12, 13, 14]:
        raise HTTPException(
            status_code=400,
            detail=f"Ugyldig GTIN-lengde: {len(clean_gtin)}. Må være 8, 12, 13 eller 14 siffer."
        )

    # Sjekk om GTIN finnes i Matinfo
    matinfo_match = None
    if clean_gtin:
        matinfo_result = await db.execute(
            select(MatinfoProduct).where(MatinfoMatinfoProduct.gtin == clean_gtin)
        )
        matinfo_match = matinfo_result.scalar_one_or_none()

    # Oppdater produkt
    old_gtin = produkt.ean_kode
    produkt.ean_kode = clean_gtin if clean_gtin else None

    await db.commit()
    await db.refresh(produkt)

    return {
        "produktid": produkt_id,
        "produktnavn": produkt.produktnavn,
        "old_gtin": old_gtin,
        "new_gtin": clean_gtin,
        "matinfo_match": {
            "found": matinfo_match is not None,
            "product_name": matinfo_match.name if matinfo_match else None,
            "brand": matinfo_match.brand if matinfo_match else None
        } if clean_gtin else None,
        "message": "GTIN oppdatert"
    }


@router.post("/bulk-update-gtin")
async def bulk_update_gtin(
    updates: BulkGtinUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Dict[str, Any]:
    """
    Masse-oppdatering av GTIN for flere produkter.

    Returnerer statistikk over oppdateringer og eventuelle feil.
    """
    results = {
        "total": len(updates.updates),
        "success": 0,
        "failed": 0,
        "matinfo_matches": 0,
        "details": []
    }

    for update in updates.updates:
        try:
            # Hent produkt
            result = await db.execute(
                select(ProdukterModel).where(ProdukterModel.produktid == update.produktid)
            )
            produkt = result.scalar_one_or_none()

            if not produkt:
                results["failed"] += 1
                results["details"].append({
                    "produktid": update.produktid,
                    "status": "error",
                    "message": "Produkt ikke funnet"
                })
                continue

            # Rens GTIN
            clean_gtin = update.gtin.strip().replace("-", "").replace(" ", "")

            # Valider lengde
            if clean_gtin and len(clean_gtin) not in [8, 12, 13, 14]:
                results["failed"] += 1
                results["details"].append({
                    "produktid": update.produktid,
                    "status": "error",
                    "message": f"Ugyldig GTIN-lengde: {len(clean_gtin)}"
                })
                continue

            # Sjekk Matinfo
            matinfo_match = None
            if clean_gtin:
                matinfo_result = await db.execute(
                    select(MatinfoProduct).where(MatinfoMatinfoProduct.gtin == clean_gtin)
                )
                matinfo_match = matinfo_result.scalar_one_or_none()
                if matinfo_match:
                    results["matinfo_matches"] += 1

            # Oppdater
            old_gtin = produkt.ean_kode
            produkt.ean_kode = clean_gtin if clean_gtin else None

            results["success"] += 1
            results["details"].append({
                "produktid": update.produktid,
                "produktnavn": produkt.produktnavn,
                "status": "success",
                "old_gtin": old_gtin,
                "new_gtin": clean_gtin,
                "matinfo_match": matinfo_match is not None
            })

        except Exception as e:
            results["failed"] += 1
            results["details"].append({
                "produktid": update.produktid,
                "status": "error",
                "message": str(e)
            })

    # Commit alle endringer
    await db.commit()

    return results


@router.get("/{produkt_id}/matinfo-suggestions", response_model=List[MatinfoSearchResult])
async def get_matinfo_suggestions(
    produkt_id: int,
    limit: int = Query(10, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> List[MatinfoSearchResult]:
    """
    Få Matinfo-forslag basert på produktnavn.

    Bruker fuzzy matching på produktnavn for å finne lignende produkter i Matinfo.
    """
    # Hent produkt
    result = await db.execute(
        select(ProdukterModel).where(ProdukterModel.produktid == produkt_id)
    )
    produkt = result.scalar_one_or_none()

    if not produkt:
        raise HTTPException(status_code=404, detail="Produkt ikke funnet")

    # Bruk produktnavn for søk
    search_term = produkt.produktnavn.strip().lower()

    # Fjern vanlige prefikser/suffikser som kan forstyrre søket
    search_term = search_term.replace("kg", "").replace("stk", "").replace("gram", "").strip()

    sql = text("""
        SELECT
            id,
            GREATEST(
                similarity(LOWER(name), :search),
                similarity(LOWER(COALESCE(brand_name, '')), :search)
            ) as similarity_score
        FROM matinfo_products
        WHERE
            LOWER(name) % :search
            OR LOWER(name) LIKE :like_search
            OR LOWER(COALESCE(brand_name, '')) % :search
        ORDER BY similarity_score DESC
        LIMIT :limit
    """)

    result = await db.execute(
        sql,
        {
            "search": search_term,
            "like_search": f"%{search_term}%",
            "limit": limit
        }
    )

    rows = result.fetchall()
    product_ids = [row.id for row in rows]
    similarity_map = {row.id: row.similarity_score for row in rows}

    # Hent produkter med allergen og nutrition data
    products_result = await db.execute(
        select(MatinfoProduct)
        .options(selectinload(MatinfoProduct.allergens), selectinload(MatinfoProduct.nutrients))
        .where(MatinfoMatinfoProduct.id.in_(product_ids))
    )
    products = products_result.scalars().all()

    # Bygg respons med sortering basert på similarity
    products_map = {p.id: p for p in products}

    return [
        MatinfoSearchResult(
            id=pid,
            gtin=products_map[pid].gtin,
            name=products_map[pid].name,
            brand=products_map[pid].brand_name,
            ingredients=products_map[pid].ingredient_statement,
            similarity_score=round(float(similarity_map[pid]), 3) if similarity_map[pid] else None,
            allergens=[
                MatinfoAllergenInfo(
                    code=a.code,
                    name=a.name,
                    level=a.level
                ) for a in products_map[pid].allergens
            ],
            nutrients=[
                MatinfoNutrientInfo(
                    code=n.code,
                    name=n.name,
                    measurement=float(n.measurement) if n.measurement else None,
                    measurement_precision=n.measurement_precision,
                    measurement_type=n.measurement_type
                ) for n in products_map[pid].nutrients
            ]
        )
        for pid in product_ids if pid in products_map
    ]

File: frontend/src/lib/api/produkter.ts
import { apiClient } from '@/lib/api-client'

export interface Produkt {
  id: number  // Alias for produktid to satisfy CrudItem constraint
  produktid: number
  produktnavn: string
  leverandorsproduktnr: string | null
  antalleht: number | null
  pakningstype: string | null
  pakningsstorrelse: string | null
  pris: number | null
  paknpris: number | null
  levrandorid: number | null
  kategoriid: number | null
  lagermengde: number | null
  bestillingsgrense: number | null
  bestillingsmengde: number | null
  ean_kode: string | null
  utgatt: boolean | null
  oppdatert: string | null
  webshop: boolean | null
  rett_komponent: boolean | null
  mvaverdi: number | null
  lagerid: number | null
  utregningsfaktor: number | null
  utregnetpris: number | null
  visningsnavn: string | null
  visningsnavn2: string | null
}

export interface ProduktListParams {
  skip?: number
  limit?: number
  aktiv?: boolean
  kategori?: number
  sok?: string
  has_gtin?: boolean  // Filter for GTIN-status: true=has GTIN, false=missing GTIN
  sort_by?: string    // Sort by field (produktid, produktnavn, pris, ean_kode)
  sort_order?: 'asc' | 'desc'  // Sort order
  include_stats?: boolean  // Include GTIN statistics (avoids N+1 queries for stats)
}

export interface ProduktStats {
  total: number
  with_gtin: number
  without_gtin: number
}

export interface ProduktListResponse {
  items: Produkt[]
  total: number
  page: number
  page_size: number
  total_pages: number
  stats?: ProduktStats  // Optional stats when include_stats=true
}

export interface MatinfoAllergenInfo {
  code: string
  name: string
  level: string  // FREE_FROM, CONTAINS, MAY_CONTAIN
}

export interface MatinfoNutrientInfo {
  code: string
  name: string
  measurement: number | null
  measurement_precision: string | null
  measurement_type: string | null
}

export interface MatinfoSearchResult {
  id: number
  gtin: string
  name: string
  brand: string | null
  ingredients: string | null
  similarity_score: number | null
  allergens: MatinfoAllergenInfo[]
  nutrients: MatinfoNutrientInfo[]
}

export interface GtinUpdateResponse {
  produktid: number
  produktnavn: string
  old_gtin: string | null
  new_gtin: string
  matinfo_match: {
    found: boolean
    product_name: string | null
    brand: string | null
  } | null
  message: string
}

export interface BulkGtinUpdate {
  produktid: number
  gtin: string
}

export interface BulkGtinUpdateResponse {
  total: number
  success: number
  failed: number
  matinfo_matches: number
  details: Array<{
    produktid: number
    produktnavn?: string
    status: 'success' | 'error'
    old_gtin?: string | null
    new_gtin?: string
    matinfo_match?: boolean
    message?: string
  }>
}

export type ProduktCreateData = {
  produktnavn: string
  leverandorsproduktnr?: string | null
  antalleht?: number | null
  pakningstype?: string | null
  pakningsstorrelse?: string | null
  pris?: number | null
  paknpris?: number | null
  levrandorid?: number | null
  kategoriid?: number | null
  lagermengde?: number | null
  bestillingsgrense?: number | null
  bestillingsmengde?: number | null
  ean_kode?: string | null
  utgatt?: boolean | null
  webshop?: boolean | null
  mvaverdi?: number | null
  lagerid?: number | null
  utregningsfaktor?: number | null
  utregnetpris?: number | null
  visningsnavn?: string | null
  visningsnavn2?: string | null
  rett_komponent?: boolean | null
}

export const produkterApi = {
  /**
   * List produkter with pagination and filtering
   */
  async list(params?: ProduktListParams): Promise<Produkt[]> {
    const response = await this.listWithMeta(params)
    return response.items || []
  },

  /**
   * List produkter with pagination and filtering, returning full response with metadata
   */
  async listWithMeta(params?: ProduktListParams): Promise<ProduktListResponse> {
    const queryParams = new URLSearchParams()

    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString())
    if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString())
    if (params?.aktiv !== undefined) queryParams.append('aktiv', params.aktiv.toString())
    if (params?.kategori !== undefined) queryParams.append('kategori', params.kategori.toString())
    if (params?.sok) queryParams.append('sok', params.sok)
    if (params?.has_gtin !== undefined) queryParams.append('has_gtin', params.has_gtin.toString())
    if (params?.sort_by) queryParams.append('sort_by', params.sort_by)
    if (params?.sort_order) queryParams.append('sort_order', params.sort_order)
    if (params?.include_stats) queryParams.append('include_stats', 'true')

    const response = await apiClient.get<ProduktListResponse>(`/v1/produkter/?${queryParams.toString()}`)
    return response.data
  },

  /**
   * Get a single produkt by ID
   */
  async get(id: number): Promise<Produkt> {
    const response = await apiClient.get<Produkt>(`/v1/produkter/${id}`)
    return response.data
  },

  /**
   * Create a new produkt
   */
  async create(data: ProduktCreateData): Promise<Produkt> {
    const response = await apiClient.post<Produkt>('/v1/produkter/', data)
    return response.data
  },

  /**
   * Update produkt
   */
  async update(id: number, data: Partial<Produkt>): Promise<Produkt> {
    const response = await apiClient.put<Produkt>(`/v1/produkter/${id}`, data)
    return response.data
  },

  /**
   * Delete produkt
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(`/v1/produkter/${id}`)
  },

  /**
   * Search Matinfo products with fuzzy matching
   */
  async searchMatinfo(query: string, limit: number = 20): Promise<MatinfoSearchResult[]> {
    const response = await apiClient.get<MatinfoSearchResult[]>(
      `/v1/produkter/matinfo/search?query=${encodeURIComponent(query)}&limit=${limit}`
    )
    return response.data
  },

  /**
   * Get Matinfo suggestions for a produkt based on its name
   */
  async getMatinfoSuggestions(produktId: number, limit: number = 10): Promise<MatinfoSearchResult[]> {
    const response = await apiClient.get<MatinfoSearchResult[]>(
      `/v1/produkter/${produktId}/matinfo-suggestions?limit=${limit}`
    )
    return response.data
  },

  /**
   * Update GTIN for a single produkt
   */
  async updateGtin(produktId: number, gtin: string): Promise<GtinUpdateResponse> {
    const response = await apiClient.patch<GtinUpdateResponse>(
      `/v1/produkter/${produktId}/gtin`,
      { gtin }
    )
    return response.data
  },

  /**
   * Bulk update GTIN for multiple produkter
   */
  async bulkUpdateGtin(updates: BulkGtinUpdate[]): Promise<BulkGtinUpdateResponse> {
    const response = await apiClient.post<BulkGtinUpdateResponse>(
      `/v1/produkter/bulk-update-gtin`,
      { updates }
    )
    return response.data
  },
}
